dask/array/tests/test_array_core.py

```diff
@@ -1541,6 +1541,62 @@ def func(dtype, block_info=None):
     assert_eq(x, np.arange(8, dtype=np.float32))
 
 
+def test_map_blocks_unique_name_chunks_dtype():
+    def func(block_info=None):
+        loc = block_info[None]["array-location"]
+        dtype = block_info[None]["dtype"]
+        return np.arange(loc[0][0], loc[0][1], dtype=dtype)
+
+    x = da.map_blocks(func, chunks=((5, 3),), dtype=np.float32)
+    assert x.chunks == ((5, 3),)
+    assert_eq(x, np.arange(8, dtype=np.float32))
+
+    y = da.map_blocks(func, chunks=((2, 2, 1, 3),), dtype=np.float32)
+    assert y.chunks == ((2, 2, 1, 3),)
+    assert_eq(y, np.arange(8, dtype=np.float32))
+    assert x.name != y.name
+
+    z = da.map_blocks(func, chunks=((5, 3),), dtype=np.float64)
+    assert z.chunks == ((5, 3),)
+    assert_eq(z, np.arange(8, dtype=np.float64))
+    assert x.name != z.name
+    assert y.name != z.name
+
+
+def test_map_blocks_unique_name_drop_axis():
+    def func(some_3d, block_info=None):
+        if not block_info:
+            return some_3d
+        return np.zeros(block_info[None]["shape"])
+
+    input_arr = da.zeros((3, 4, 5), chunks=((3,), (4,), (5,)), dtype=np.float32)
+    x = da.map_blocks(func, input_arr, drop_axis=[0], dtype=np.float32)
+    assert x.chunks == ((4,), (5,))
+    assert x.compute().shape == (4, 5)
```

Comment:
Nitpick: In this test and `test_map_blocks_unique_name_new_axis`, we're explicitly computing and checking the shape of the output array, but in `test_map_blocks_unique_name_chunks_dtype` we're using `assert_eq` which does this shape check along with additional sanity checks. I'm wondering if there's a reason to just check the computed `shape` is as expected, or if we could use `assert_eq` in these tests too. 

For example, replacing 

```python
assert x.compute().shape == (4, 5)
```

with something like

```python
assert_eq(x, np.zeros((4, 5), dtype=np.float32))
```

That would ensure other checks (e.g. dtype is as expected) are also run and would be more consistent with other tests generally

Reply:
As long as `assert_eq` is actually computing the data and *then* checking the shape, then I will change that. Otherwise it is using the `.shape` of the dask Array which may be incorrect if the map_blocks'd function is behaving badly.

Reply:
Done. And good call too because the dtype was wrong :wink:

Reply:
That's a good distinction to make. One of the checks in `assert_eq(a, b)` is that the `compute`d shape of both `a` and `b` are equal, so it should provide coverage for what you had initially